ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt

```diff
@@ -33,7 +35,11 @@ public sealed class HandlerDispatcher : CoroutineDispatcher(), Delay {
      * **Note: This is an experimental api.** Semantics of this dispatcher may change in the future.
      */
     @ExperimentalCoroutinesApi
-    public abstract val immediate: HandlerDispatcher
+    public abstract override val immediate: HandlerDispatcher
+}
+
+internal class AndroidDispatcherFactory : MainDispatcherFactory {
```

Comment:
This class needs annotated with `@Keep` or it will be removed automatically as being unused for all R8/ProGuard users requiring manual configuration.

Reply:
Good catch, thank you.

>I suspect this might cause start performance regression for all Android users based on putting ServiceLoader usage on the critical path

`mainDispatcher` is initialized only once when `Dispatchers` object is instantiated, it is not on the critical path.

Reply:
I suspect most Android apps will launch _something_ during startup which requires a `Dispatchers.Main` reference and will force initialization.

Perhaps we can get R8 to rewrite any `ServiceLoader` usage with a constant array determined at compile-time. This could avoid the need for `@Keep` on the implementations as well.

Reply:
I got your point. Is an access to `ServiceLoader` a regression significant enough to postpone this change?

Reply:
I don't know what the impact is off-hand. Android definitely isn't optimized for this case though which is why it has its own resource and asset system. I can measure, but it won't be until Friday at the earliest.

Reply:
We will merge it then, mostly to stabilize our API.
If you (or anyone else) report a significant regression, we will rework this part without changing the API (e.g. with `Class.forName` or any other hack like this)